corehq/apps/locations/tasks.py

```diff
@@ -174,13 +176,39 @@ def import_locations_async(domain, file_ref_id):
 
 
 @task
-def update_users_at_locations(location_ids):
+def update_users_at_locations(domain, location_ids, supply_point_ids, ancestor_ids):
     """
     Update location fixtures for users given locations
     """
-    from corehq.apps.users.models import update_fixture_status_for_users
+    from corehq.apps.users.models import CommCareUser, update_fixture_status_for_users
     from corehq.apps.locations.dbaccessors import user_ids_at_locations
     from corehq.apps.fixtures.models import UserFixtureType
-
-    user_ids = user_ids_at_locations(location_ids)
+    from dimagi.utils.couch.database import iter_docs
+
+    def has_ref(obj, location_id):
+        # obj is either a CommCareUser or DomainMembership object
+        return (
+            obj.location_id == location_id or
+            location_id in obj.assigned_location_ids
+        )
+
+    # close supply point cases
+    for supply_point_id in supply_point_ids:
+        close_supply_point_case(domain, supply_point_id)
+
+    # unassign users from locations
+    unassign_user_ids = user_ids_at_locations(location_ids)
+    for doc in iter_docs(CommCareUser.get_db(), unassign_user_ids):
+        user = CommCareUser.wrap(doc)
+        membership = user.get_domain_membership(domain) if user.is_web_user() else user
+        for location_id in location_ids:
+            if not has_ref(membership, location_id):
+                continue
```

Comment:
Could this `has_ref` check be moved to the `unset_location_by_id` methods?  Seems like they should be responsible for noop-ing when appropriate.  Looks like they do have some checking around that.